### ISSUE TITLE
Every parameter is mut

### ISSUE THREAD

**Issue by gustavla** [stainless-steel/lapack-sys#2]
I've noticed that every parameter is specified as mutable, even the ones that don't need to be. This is making calls a lot more awkward, since all the variables now have to be mutable. This is not the case in blas-sys, so I was wondering why it was the case here.

I might open a pull request addressing this, unless there is a reason for it.

I'm using blas-sys and lapack-sys in a project I call [numeric](https://github.com/gustavla/numeric) and it's working out great so far, except interfacing with lapack is a lot messier than blas because of this issue ([example](https://github.com/gustavla/numeric/blob/4c0d323dd6452953ad08221faae2c68f5fd5dd0b/src/linalg/mod.rs)).

**Comment by IvanUkhov** [comment]
@gustavla, thanks for your feedback. Initially, apart from the Fortran interfaces, both `blas-sys` and `lapack-sys` also included the corresponding C interfaces, CBLAS and LAPACKE. When deciding on Rust definitions in the `-sys` crates, the idea was to match exactly/blindly the signatures of the corresponding C functions as it was defined by Netlib. Higher level interfaces, [blas](https://github.com/stainless-steel/blas) and [lapack](https://github.com/stainless-steel/lapack), were supposed to take take of that. So we had sort of [role](https://raw.githubusercontent.com/copies/lapack/master/CBLAS/include/cblas.h) [models](https://raw.githubusercontent.com/copies/lapack/master/LAPACKE/include/lapacke.h) for `blas-sys` and `lapack-sys`, which we tried to follow; the mutability in the LAPACK part was a consequence of that.

I agree that in many cases mutable arguments don’t make any sense; they’re mutable but they shouldn’t really be. It certainly would be great if somebody could fix that. Unless one finds an automatic way of doing this, it would probably require a lot of patience and hard work. This is one of the reason that [lapack](https://github.com/stainless-steel/lapack) is incomplete; [blas](https://github.com/stainless-steel/blas), on the other hand, is complete and is meant to be used instead of `blas-sys`.

In any case, I would appreciate any help. Thanks!


Regards,
Ivan

**Comment by gustavla** [comment]
What you are saying is essentially that time is better spent on the lapack project, making neat rust interfaces for these functions? I might port my stuff over to blas/lapack (instead of *-sys) then and open pull requests to improve lapack if functions I need are missing.

**Comment by IvanUkhov** [comment]
@gustavla, I think that eliminating the excessive mutability in `lapack-sys` is a very healthy thing to do: there is no any particular reason to adhere to some poor C interface such for the sake of it. Moreover, as you noted yourself, higher level interfaces suffer from this quite a bit. I’d have already got rid of this mutability if I had found a clean, automatic, error-free way of doing this.

To answer your question directly: Yes, I believe that, if you have time, it would be better to spend it on the `lapack` crate. The way we did it for the `blas` crate was a [handy Python script](https://github.com/stainless-steel/blas/blob/master/etc/gen.py), which could probably be useful for `lapack` as well with some adjustments.

Thanks!


Regards,
Ivan
